diff --git a/src/frame.c b/src/frame.c
index cb80ba24a1c..dd3ae96c98c 100644
--- a/src/frame.c
+++ b/src/frame.c
@@ -971,6 +971,7 @@ make_frame (bool mini_p)
   f->want_fullscreen = FULLSCREEN_NONE;
   f->undecorated = false;
   f->undecorated_round = false;
+  f->has_shadow = true;
   f->no_special_glyphs = false;
 #ifndef HAVE_NTGUI
   f->override_redirect = false;
@@ -3964,6 +3965,7 @@ static const struct frame_parm_table frame_parms[] =
   {"inhibit-double-buffering",  SYMBOL_INDEX (Qinhibit_double_buffering)},
   {"undecorated",		SYMBOL_INDEX (Qundecorated)},
   {"undecorated-round",		SYMBOL_INDEX (Qundecorated_round)},
+  {"has-shadow",		SYMBOL_INDEX (Qhas_shadow)},
   {"parent-frame",		SYMBOL_INDEX (Qparent_frame)},
   {"skip-taskbar",		SYMBOL_INDEX (Qskip_taskbar)},
   {"no-focus-on-map",		SYMBOL_INDEX (Qno_focus_on_map)},
@@ -6184,6 +6186,7 @@ syms_of_frame (void)
   DEFSYM (Qminibuffer, "minibuffer");
   DEFSYM (Qundecorated, "undecorated");
   DEFSYM (Qundecorated_round, "undecorated-round");
+  DEFSYM (Qhas_shadow, "has-shadow");
   DEFSYM (Qno_special_glyphs, "no-special-glyphs");
   DEFSYM (Qparent_frame, "parent-frame");
   DEFSYM (Qskip_taskbar, "skip-taskbar");
diff --git a/src/frame.h b/src/frame.h
index 5a2fc4e0212..fa053ee87bf 100644
--- a/src/frame.h
+++ b/src/frame.h
@@ -400,6 +400,9 @@ struct frame
   /* True if this is an undecorated frame with round corners.  */
   bool_bf undecorated_round : 1;
 
+  /* True if this is frame should have a shadow.  */
+  bool_bf has_shadow : 1;
+
 #ifndef HAVE_NTGUI
   /* True if this is an override_redirect frame.  */
   bool_bf override_redirect : 1;
@@ -1084,6 +1087,7 @@ default_pixels_per_inch_y (void)
 #if defined (HAVE_WINDOW_SYSTEM)
 #define FRAME_UNDECORATED(f) ((f)->undecorated)
 #define FRAME_UNDECORATED_ROUND(f) ((f)->undecorated_round)
+#define FRAME_HAS_SHADOW(f) ((f)->has_shadow)
 #ifdef HAVE_NTGUI
 #define FRAME_OVERRIDE_REDIRECT(f) ((void) (f), 0)
 #else
@@ -1111,6 +1115,7 @@ default_pixels_per_inch_y (void)
 #else /* not HAVE_WINDOW_SYSTEM */
 #define FRAME_UNDECORATED(f) ((void) (f), 0)
 #define FRAME_UNDECORATED_ROUND(f) ((void) (f), 0)
+#define FRAME_HAS_SHADOW(f) ((void) (f), 1)
 #define FRAME_OVERRIDE_REDIRECT(f) ((void) (f), 0)
 #define FRAME_PARENT_FRAME(f) ((void) (f), NULL)
 #define FRAME_SKIP_TASKBAR(f) ((void) (f), 0)
diff --git a/src/nsfns.m b/src/nsfns.m
index f14ac7ac7cc..567ca8b13ea 100644
--- a/src/nsfns.m
+++ b/src/nsfns.m
@@ -1012,6 +1012,28 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
     error ("Both left and top icon corners of icon must be specified");
 }
 
+void
+ns_set_has_shadow (struct frame *f, Lisp_Object new_value, Lisp_Object old_value)
+/* --------------------------------------------------------------------------
+     Set frame F's `has-shadow' parameter. If non-nil, F's window-system
+     window is drawn with a shadow. If nil, the window has no shadow.
+   -------------------------------------------------------------------------- */
+{
+  NSTRACE ("ns_set_has_shadow");
+
+  if (!EQ (new_value, old_value))
+    {
+      block_input ();
+
+      FRAME_HAS_SHADOW (f) = !NILP (new_value);
+
+      if (FRAME_NS_VIEW (f))
+        [(EmacsWindow *)[FRAME_NS_VIEW (f) window] setHasShadow: FRAME_HAS_SHADOW (f)];
+
+      unblock_input ();
+    }
+}
+
 
 /* Note: see frame.c for template, also where generic functions are
    implemented.  */
@@ -1059,6 +1081,7 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
   0, /* x_set_inhibit_double_buffering */
   ns_set_undecorated,
   ns_set_undecorated_round,
+  ns_set_has_shadow,
   ns_set_parent_frame,
   0, /* x_set_skip_taskbar */
   ns_set_no_focus_on_map,
@@ -1391,6 +1414,12 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
   FRAME_UNDECORATED_ROUND (f) = !NILP (tem) && !EQ (tem, Qunbound);
   store_frame_param (f, Qundecorated_round, FRAME_UNDECORATED_ROUND (f) ? Qt : Qnil);
 
+  tem = gui_display_get_arg (dpyinfo, parms, Qhas_shadow, NULL, NULL,
+                             RES_TYPE_BOOLEAN);
+  FRAME_HAS_SHADOW (f) = !NILP (tem) && !EQ (tem, Qunbound);
+  store_frame_param (f, Qhas_shadow, FRAME_HAS_SHADOW (f) ? Qt : Qnil);
+
+
 #ifdef NS_IMPL_COCOA
 #ifndef NSAppKitVersionNumber10_14
 #define NSAppKitVersionNumber10_14 1671
diff --git a/src/nsterm.h b/src/nsterm.h
index 317093a65ce..1394f547c66 100644
--- a/src/nsterm.h
+++ b/src/nsterm.h
@@ -1229,6 +1229,8 @@ extern void ns_set_undecorated (struct frame *f, Lisp_Object new_value,
                                 Lisp_Object old_value);
 extern void ns_set_undecorated_round (struct frame *f, Lisp_Object new_value,
                                       Lisp_Object old_value);
+extern void ns_set_has_shadow (struct frame *f, Lisp_Object new_value,
+                                      Lisp_Object old_value);
 extern void ns_set_parent_frame (struct frame *f, Lisp_Object new_value,
                                  Lisp_Object old_value);
 extern void ns_set_no_focus_on_map (struct frame *f, Lisp_Object new_value,
diff --git a/src/nsterm.m b/src/nsterm.m
index 05b6e1029f9..05ec3902484 100644
--- a/src/nsterm.m
+++ b/src/nsterm.m
@@ -9428,14 +9428,8 @@ - (void)setParentChildRelationships
   [ourView updateCollectionBehavior];
 #endif
 
-  /* Child frames are often used in ways that may mean they should
-     "disappear" into the contents of the parent frame.  macOs's
-     drop-shadows break this effect, so remove them on undecorated
-     child frames.  */
-  if (parentFrame && FRAME_UNDECORATED (ourFrame))
-    [self setHasShadow:NO];
-  else
-    [self setHasShadow:YES];
+  /* Use the frame's has-shadow parameter */
+  [self setHasShadow: FRAME_HAS_SHADOW (ourFrame)];
 #endif
 
 
